app/services/spell_check_service.rb

```
@@ -0,0 +1,90 @@
+class SpellCheckService
+  SpellCheckServiceError = Class.new(StandardError)
+  SpellCheckBingAPIError = Class.new(SpellCheckServiceError)
+  BadRequestError = Class.new(SpellCheckBingAPIError)
+  UnauthorizedError = Class.new(SpellCheckBingAPIError)
+  ForbiddenError = Class.new(SpellCheckBingAPIError)
+  RateLimitExceededError = Class.new(SpellCheckBingAPIError)
+
+  API_ENDPOINT = 'https://api.cognitive.microsoft.com/bing/v7.0/spellcheck'.freeze
+
+  HTTP_OK_CODE = 200
+  HTTP_BAD_REQUEST_CODE = 400
+  HTTP_UNAUTHORIZED_CODE = 401
+  HTTP_FORBIDDEN_CODE = 403
+  HTTP_RATE_LIMIT_EXCEEDED_CODE = 429
+
+  attr_reader :response
+
+  def initialize(api_key: Rails.configuration.bing_spell_check_api_key)
+    @api_key = api_key
+  end
+
+  def scan(search_term: nil)
+    return unless search_term
+
+    service_enabled?
+
+    bing_spell_check(search_term: search_term)
+  rescue RuntimeError => e
+    Rails.logger.error("Spell Check service error: #{e.message}")
+    raise SpellCheckServiceError, e.message
+  end
+
+  private
+
+  def bing_spell_check(search_term: nil)
+    @response = client.get(API_ENDPOINT) { |req|
+      req.params['mkt'] = 'en-gb'
+      req.params['mode'] = 'spell'
+      req.params['text'] = search_term
+    }
+
+    return suggested_correction if response_successful?
+
+    handle_api_errors
+  end
+
+  def suggested_correction
+    JSON.parse(response.body)['flaggedTokens'].map { |token| token['suggestions'].first['suggestion'] }.join(' ')
+  end
+
+  def client
+    @client ||= Faraday.new { |client|
+      client.request :url_encoded
+      client.adapter Faraday.default_adapter
+      client.headers['Content-Type'] = 'application/x-www-form-urlencoded'
+      client.headers['Ocp-Apim-Subscription-Key'] = api_key
+    }
+  end
+
+  def api_key
+    raise 'BingSpellCheck API Key is not set' unless @api_key.present?
```

I don't htink we need to raise an exception when there is  no key available, but rather fail silently. As this makes development and tests fail otherwise when they touch this page

---

Hmm.. I think we have this pattern in a couple of other services already, but the expected behaviour for this service is subtly different. It's essentially an optional service, so if no API key is defined then the service shouldn't be invoked. I'm not sure where the best place for that logic would be, however with the current implementation this would mean the controller needs to rescue from this exception and presumably ignore it. It's generally not recommended to raise an exception by design for an expected case - they should be used (as the name suggests) for exceptional scenarios).

---

I think it's more useful to see a clear reason rather than seeing nil. Besides we have the same pattern for Notify.

---

Sorry I don’t agree. Notify is a dependency whereas this is optional functionality - it won’t be configured with an API key locally or on dev environments so doesn’t need to raise exceptions.

---

In which case the feature will be off. We can't have the service ON with no API key. What's the point?